### PR TITLE
Update beaker setup dependencies with linchpin

### PR DESCRIPTION
**Pull request by samvarankashyap**
Install beaker dependencies from beaker repo

### CODE REVIEW

**Comment by samvarankashyap** [comment]
[test]

File: linchpin/provision/roles/dependencies/tasks/main.yml [hunk 1]
```diff
@@ -10,11 +10,11 @@
   loop_control:
     loop_var: indv_prov
 
-- name: Install os-specific depdendencies:: Fedora
+- name: "Install os-specific depdendencies Fedora"
   include_tasks: "setup_dnf_deps.yml"
   when: setup_providers | length == 0  and ansible_distribution == 'Fedora'
 
-- name: Install os-specific depdendencies:: CentOS/RHEL
+- name: "Install os-specific depdendencies CentOS/RHEL"
```
Comment (StykMartin) [inline linchpin/provision/roles/dependencies/tasks/main.yml]: Are you sure about the quotes you added? 
Another thing you have typo there depdendencies -> dependencies

File: linchpin/provision/roles/dependencies/tasks/main.yml [hunk 2]
```diff
@@ -10,11 +10,11 @@
   loop_control:
     loop_var: indv_prov
 
-- name: Install os-specific depdendencies:: Fedora
+- name: "Install os-specific depdendencies Fedora"
```
Comment (StykMartin) [inline linchpin/provision/roles/dependencies/tasks/main.yml]: Are you sure about the quotes you added? 
Another thing you have typo there depdendencies -> dependencies

File: linchpin/provision/roles/dependencies/tasks/beaker_deps.yml [hunk 3]
```diff
@@ -1,9 +1,3 @@
 - name: Install pypi dependencies of beaker
   pip:
-    name: "{{ bkr_pkg }}"
-  with_items:
-    - 'beaker-client>=23.3'
-    - 'python-krbV'
-    - 'bottle'
-  loop_control:
-    loop_var: bkr_pkg
+    name: ['git+https://github.com/beaker-project/beaker.git@develop#subdirectory=Client/', 'python-krbV', 'bottle']
```
Comment (StykMartin) [inline linchpin/provision/roles/dependencies/tasks/beaker_deps.yml]: I can see that you are trying to get beaker from Develop branch. Develop branch contains Beaker 27.
I replaced python-krbV with python-gssapi

File: linchpin/provision/roles/dependencies/tasks/main.yml (hunk 1 above)
Comment (samvarankashyap) [reply linchpin/provision/roles/dependencies/tasks/main.yml]: Adding special characters like "::" causes failures in ansible. 
The is the reason I have added quotes to name. Else it is not necessary. 
I will fix typo

File: linchpin/provision/roles/dependencies/tasks/beaker_deps.yml (hunk 3 above)
Comment (samvarankashyap) [reply linchpin/provision/roles/dependencies/tasks/beaker_deps.yml]: I could not find a tag specific to beaker client that is the reason I have used to develop branch code. 
Further, does the above mean python-krbV is no longer needed?

File: linchpin/provision/roles/dependencies/tasks/main.yml (hunk 1 above)
Comment (StykMartin) [reply linchpin/provision/roles/dependencies/tasks/main.yml]: Sure, I understand that.
But you removed "::" from the name, therefore you don't need quotes.

1. Remove "::" and you don't need quotes.
2. Keep "::" and add quotes.

And typo is not fixed yet.

File: linchpin/provision/roles/dependencies/tasks/beaker_deps.yml (hunk 3 above)
Comment (StykMartin) [reply linchpin/provision/roles/dependencies/tasks/beaker_deps.yml]: Beaker client doesn't require python-krbV anymore. python-gssapi is sufficient for beaker client / common.

Comment (StykMartin) [reply linchpin/provision/roles/dependencies/tasks/beaker_deps.yml]: + I think you should consume our tags.. Not blindly install develop

**Comment by samvarankashyap** [comment]
@14rcole Needs review

File: linchpin/provision/roles/dependencies/tasks/beaker_deps.yml (hunk 3 above)
Comment (samvarankashyap) [reply linchpin/provision/roles/dependencies/tasks/beaker_deps.yml]: I completely agree with you. 
However, could share the link/tag which contains python3 patch
the GitHub repo seems confusing. 
Thank you.

**Comment by Dannyb48** [comment]
Can we push this out to 1.9.0 release? I ran into an issue manually testing the beaker-client on github locally in a virtualenv. The gssapi caused issues with the paramiko version I was using. I just want to spend time from the carbon side make sure this works.

File: linchpin/provision/roles/dependencies/tasks/beaker_deps.yml (hunk 3 above)
Comment (StykMartin) [reply linchpin/provision/roles/dependencies/tasks/beaker_deps.yml]: Hello @samvarankashyap.
Python3 patch is not tagged (yet). I'm working on it, we have to finish Beaker 27 changes. Then I can provide you tag. I'm about to release 26.6. And we are close to 27 - We are performing tests in stage.

File: linchpin/provision/roles/dependencies/tasks/main.yml (hunk 1 above)
Comment (ernestask) [reply linchpin/provision/roles/dependencies/tasks/main.yml]: Why even have the double colon?

**Comment by StykMartin** [comment]
Are you in hurry with this patch?
When Beaker 27.0 will be released I will also push changes to currently unmaintained PyPI.

I think I can help you guys to make this clean and don't install mess into client machines

**Comment by samvarankashyap** [comment]
@StykMartin : 
We are planning to add the patch for the next release which is due October 28th. 
quick question is there any tentative date for beaker 27 release? 
Thank you

**Comment by StykMartin** [comment]
@samvarankashyap :
Our release is blocked right now by testing performed in a staging environment.
I hope we can manage it within a month. But hard to say.
Right now one team is doing testing and they didn't provide a new status. I will try to reach them if they can provide a date.

**Comment by Dannyb48** [comment]
I think we should at least test using develop branch or the internal dev rpm to test out the Linchpin Beaker provisioning. To at least make sure the Linchpin python 3 compatible side works consuming the python beaker 3 client apis? That way we can catch things early before beaker 27 is tagged and released?

**Comment by StykMartin** [comment]
For your testing purposes definitely yes! If you want to use internal dev RPM feel free to reach me out. I'm creating new RPM for every new commit merged to develop branch

**Comment by StykMartin** [comment]
Successfully it does not

**Comment by Dannyb48** [comment]
so far my testing with the internal dev beaker rpm on RHEL 8 (Red Hat Enterprise Linux release 8.0 (Ootpa)) has been successful using linchpin 1.8.2 and carbon 1.4.1. 

The minimal bkr parameters I passed in my scenario 
`
```yaml
      arch: x86_64
      distro: RHEL-7.6
      variant: Server
      whiteboard: {{ BKR_WHITEBOARD }}
      jobgroup: {{ BKR_JOBGROUP }}
      username: {{ BKR_USERNAME }}
      password: {{ BKR_PASSWORD }}
      ssh_key: keys/{{ BKR_KEYPAIR }}
      host_requires_options: 
        - "force = {{ BKR_HOST_FQDN }}"
````

The only issue I ran into was that the carbon framework had an issue with GSSAPI but that was because of we were using an older version of paramiko, 2.4.2, had to upgrade paramiko 2.6.0 which has full support for GSSAPI. 

I will test with the develop branch of linchpin next. Not I have been driving this from carbon not Linchpin CLI. 

@StykMartin I think you've mentioned in the beaker-user list that if I have a RHEL 7 w/python3 and I try to install the beaker27 client rpm, this isn't supported? or won't work?

**Comment by StykMartin** [comment]
@Dannyb48 at this moment RHEL 7 client is built for Python 2 only.
We are not providing python 3 support (yet) in RHEL 7.

**Comment by Dannyb48** [comment]
@StykMartin Sounds good thanks!

**Comment by Dannyb48** [comment]
@samvarankashyap 
My intial testing of both linchpin 1.8.2 and linchpin develop on RHEL8 and Fedora30 with the beaker common/clients 27 build 107 rpm from the internal beakerrepo download site have proven successful to provision and destroy a single beaker resource through carbon using the simple pinfile generated by carbon below.

The only real gotcha was from carbon side, where we had to update paramiko 2.4.2 to paramiko 2.6.0 to support the new beaker GSSAPI requirement. 

I did have issues trying to install beaker from the github Client sub-directory issues with finding python dependencies. 

Although I do question, based on Martin's answer, if we need to add another task to this playbook? One that installs the bkr client 27 install on systems that support it, i.e. RHEL8/Fedora30 and leave this existing task for systems <RHEL8/Fedora30?

``` yaml
carbon:
  layout:
    inventory_layout:
      hosts:
        node:
          count: 1
          host_groups:
          - hosts
          - client
      vars:
        ansible_ssh_private_key_file: keys/demo
        ansible_user: root
        hostname: db2-carbon-beaker-node
  topology:
    resource_groups:
    - resource_definitions:
      - job_group: ci-ops-central
        recipesets:
        - arch: x86_64
          count: 1
          distro: RHEL-7.6
          hostrequires:
          - force: dell-per220-02.cci.lab.eng.bos.redhat.com
          name: db2-carbon-beaker-node
          ssh_key_file:
          - demo.pub
          variant: Server
        role: bkr_server
        ssh_keys_path: /home/fedora/cbn/e2e-acceptance-tests/keys
        whiteboard: py3-beaker-testing
      resource_group_name: carbon
      resource_group_type: beaker
    topology_name: carbon
```

**Comment by samvarankashyap** [comment]
@Dannyb48 
I agree with you. 
We can maintain multiple tasks based on operating systems. 
Currently, we do test on fedora30, centos7 and fedora29. 
I will update the task accordingl

**Comment by samvarankashyap** [comment]
@StykMartin 
It would be great if we can get a tentative release date for the beaker 27 release.

**Comment by StykMartin** [comment]
I think in 3 weeks we have maintenance window to push updates.
So I'm expecting around this time we will push 27.

**Comment by Dannyb48** [comment]
FYI on RHEL 8 - when using the install from github method, we need to make sure that the `python3-gssapi` rpm bindings were installed before the python `gssapi` package would install.

Did not have this problem when the beaker-client was installed using the rpm method.

**Comment by samvarankashyap** [comment]
@Dannyb48 
I am planning to add rhel7 rhel8 support in upcoming release of linchpin 
ref: #1197 #1445 
I will keep that in mind before I do that. 
Thanks.

**Comment by StykMartin** [comment]
Working on release notes. This week we will tag RC

**Comment by samvarankashyap** [comment]
@StykMartin 
I see there is a beaker-client rpm on fedora 30 repos. 
Are there any plans pushing beaker-client to PyPI anytime soon?

**Comment by StykMartin** [comment]
@samvarankashyap beaker-client 27.0
Screenshot_20191204-182629_Chrome

**Comment by samvarankashyap** [comment]
@StykMartin Awesome news. 
Thank you very much for the update :+1:

**Comment by StykMartin** [comment]
:tada: :tada: :tada:
Sorry for prolonging this release. We had many problems to solve.

File: linchpin/provision/roles/dependencies/tasks/beaker_deps.yml [hunk 4]
```diff
@@ -1,9 +1,3 @@
 - name: Install pypi dependencies of beaker
   pip:
-    name: "{{ bkr_pkg }}"
-  with_items:
-    - 'beaker-client>=23.3'
-    - 'python-krbV'
-    - 'bottle'
-  loop_control:
-    loop_var: bkr_pkg
+    name: ['git+https://github.com/beaker-project/beaker.git@develop#subdirectory=Client/', 'python-gssapi', 'bottle']
```
Comment (StykMartin) [inline linchpin/provision/roles/dependencies/tasks/beaker_deps.yml]: This can be replaced with the latest 27.0 Beaker release.

File: setup.py [hunk 5]
```diff
@@ -46,7 +46,7 @@
     tests_require=["pytest<=4.4.0", "nose", "mock", "coverage", "flake8"],
     extras_require={
         'krbV': ["python-krbV"],
-        'beaker': ['beaker-client>=23.3', 'python-krbV'],
+        'beaker': ['beaker-client>=27.0', 'python-gssapi', 'bottle'],
```
Comment (StykMartin) [inline setup.py]: Why do you need `bottle`?
It is definitely not part of [Beaker](https://github.com/beaker-project/beaker) requirements.

**Comment by samvarankashyap** [comment]
[merge]

**Pull request merged**